First draft

tpoisot opened the pull request:
This defines one abstract type (`NeutralLandscapeMaker`), the `PlanarGradient`, `EdgeGradient`, and `NoGradient` types and algorithms, and overloads `Base.rand` and `Random.rand!` as the main method.

tpoisot committed:
✋ planar gradient

tpoisot committed:
📖 docstring

mkborregaard commented on src/algorithms/planargradient.jl:
```diff
@@ -0,0 +1,13 @@
+    eastness = sin(deg2rad(alg.direction))
+    southness = -1cos(deg2rad(alg.direction))
+    l = southness .* (1:s[1]) .+ (eastness .* (1:s[2]))'
+    return l
```
I see you moved out the masking - I agree with that. The user should be able to mask manually with a mask function, no need to have that as a keyword arg.

tpoisot replied:
so... the only issue is that the scaling is done *after* the masking, which tends to break the flow of a lot of things

tpoisot replied:
my current solution would be to mask if a mask argument is given (or a series of coordinates, maybe), and then *also* give a mask function after that

mkborregaard replied:
right, my mistake :+1:

mkborregaard commented on src/algorithms/planargradient.jl:
```diff
@@ -0,0 +1,13 @@
+
+PlanarGradient() = PlanarGradient(360rand())
+
+function landscape(alg::PlanarGradient, s::Tuple{IT,IT}=(10,30)) where {T <: Real, IT <: Integer}
```
Why is the size a tuple? Most of the Array constructors, rand etc take two separate args

tpoisot replied:
that's mostly shorthand to say "this should have two values", but that's a good point, I can change that

mkborregaard commented on src/NeutralLandscapes.jl:
```diff
@@ -2,4 +2,10 @@ module NeutralLandscapes
 
 greet() = print("Hello World!")
```
I guess we don't need greet

tpoisot replied:
forgot to remove it!

mkborregaard replied:
I took it out in #2

mkborregaard reviewed (commented):
Yes, great! This is the right appraoch imho

tpoisot committed:
✨ add EdgeGradient

Currently not working because it calls PlanarGradient, which should rescale and mask the results

gottacatchenall commented:
also the python version scales everything between 0 and 1, not sure if that should be an arg to `landscape()` or a separate call

mkborregaard commented on src/algorithms/edgegradient.jl:
```diff
@@ -0,0 +1,18 @@
+additional arguments `kw...` are passed to the post-processing function, see the
+documentation of **TODO**. 
+"""
+function landscape(alg::EdgeGradient, s::Tuple{IT,IT}=(10,30); kw...) where {IT <: Integer}
```
Could we have a function instead:
```julia
function landscape(alg, size; mask = Nothing, kw...) 
  ret = Matrix(size...)
  landscape!(alg, size; mask = mask, kw...)
end

function landscape!(mat, alg, size; mask = Nothing, kw...)
  _landscape!(mat!, alg, size; kw...)
  isnothing(mask) || mask!(ret)
  rescale!(ret)
end
```
and then have the dispatching function be `_landscape!`?

mkborregaard replied:
would save a lot of the boilerplate in the python version

mkborregaard replied:
@gottacatchenall ^

gottacatchenall replied:
i like this structure but i wonder if there is a case where a user would not want it rescaled---we could do something similar as mask, ie

```
function landscape!(mat, alg, size; scale = true, mask = Nothing, kw...)
  _landscape!(mat!, alg, size; kw...)
  isnothing(mask) || mask!(ret)
  scale && rescale!(ret)
end
```

or something similar, although i'm not sure if there is a need for that use-case

tpoisot replied:
the issue with not rescaling is that it *does* still scale, but with network size - this is not ideal, I think

tpoisot replied:
> and then have the dispatching function be `_landscape!`?

This is so obvious... Yes!

mkborregaard replied:
https://github.com/EcoJulia/NeutralLandscapes.jl/pull/2

gottacatchenall replied:
> the issue with not rescaling is that it _does_ still scale, but with network size - this is not ideal, I think

hmm i guess i'm thinking if you want different types of scaling (linear scaling on an arbitrary interval,  standardization, normalization, etc), you could use something like this to use an arbitrary scaling function

```
function landscape!(mat, alg, size; scaler = scale!, mask = Nothing, kw...)
  _landscape!(mat!, alg, size; kw...)
  isnothing(mask) || mask(ret)
  scaler(ret)
end
```

although if we want to replicate python functionality before adding new things that is good with me too

mkborregaard replied:
I'd think so - fidelity first. Scaling is also quite cheap I'd think

mkborregaard committed:
add inplace functions

mkborregaard committed:
insert NaN rather than missing

mkborregaard committed:
use NaNMath for the rescaling

mkborregaard committed:
add mask to README

mkborregaard commented:
Instead of `landscape!` and `landscape` - should it just be `rand` and `rand!`? I quite like
```julia
landscape = rand(PlanarGradient(60.0), 100, 120)
```
for its symmetry with
```julia
using Distributions
normalarray = rand(Normal(20, 3), 100, 120)
```

mkborregaard committed:
clean up rescale!

mkborregaard committed:
clean up mask!

mkborregaard committed:
move docstring for landscape

mkborregaard committed:
backticks for code

mkborregaard committed:
add plots to Readme

mkborregaard committed:
add a NoGradient type 

"random" in the python code

mkborregaard committed:
fix doctring

mkborregaard committed:
add small docstring

mkborregaard committed:
explicit using in readme

tpoisot committed:
Merge pull request #2 from EcoJulia/mkb/inplace

[RFC] add inplace functions

tpoisot committed:
📖 some docstrings

tpoisot committed:
📖 change the project

tpoisot commented:
> Instead of `landscape!` and `landscape` - should it just be `rand` and `rand!`? I quite like

I was wondering the same thing while I was cooking...  I think it makes sense.

tpoisot committed:
💥 version with rand/rand!

tpoisot committed:
fix the docstrings

tpoisot committed:
✅ tests

tpoisot committed:
✨ WaveSurface

tpoisot committed:
💥 didn't need Plots

tpoisot committed:
📖 correct the doc file

tpoisot commented:
Let's work from `main`!
